File: src/lib/presence.ts
/**
 * Who else has this project open, in one place.
 *
 * Everything here is **advisory and only advisory**. Nothing in this module,
 * and nothing that reads it, disables a control, delays a save or refuses one.
 * Hard locks over a share strand files whenever a laptop sleeps or a VPN drops
 * and the recovery is worse than the collision, so presence warns and never
 * blocks — see `docs/07-file-shares.md` and `wobu-store/src/presence.rs`.
 *
 * The wording lives here rather than in the components for the same reason
 * `readOnly.ts` exists: four surfaces say versions of one sentence — the
 * greeting on open, the navigator dot, the editor banner, the session count —
 * and they have to agree about who is here.
 */
import type { Peer } from './api'

/**
 * Seconds after which a session that has stopped writing its heartbeat is
 * treated as gone.
 *
 * The same sixty seconds as `presence::STALE_AFTER`, deliberately, and not a
 * second rule: the backend drops stale sessions when it answers, and this side
 * applies the identical bound to the answer it is holding. Two numbers here
 * would mean a peer the backend had reaped still lighting up a row, or a peer
 * who is sitting there flickering out of one.
 */
export const STALE_AFTER_SECS = 60

/**
 * How often the peer list is re-read.
 *
 * Half the heartbeat interval, so someone who arrives is on screen within about
 * a beat rather than two. Polled rather than pushed because the answer only
 * changes at human speed — see `presence_peers` for why there is no event.
 *
 * What this does not cover, honestly: polling pauses while the window is in the
 * background, so a list can be arbitrarily old on return. A refetch on focus is
 * what corrects it, and the filter below is what stops the held answer being
 * believed in the meantime.
 */
export const PRESENCE_POLL_MS = 10_000

/** One banner, keyed by code, exactly as `project.read_only` is. */
export const PRESENCE_BANNER = 'presence.editing_elsewhere'

/**
 * The peers still beating, out of whatever the last poll returned.
 *
 * `seenSecsAgo` is measured on the backend against its own clock, never against
 * the timestamp inside a peer's file — two desktops on a LAN routinely sit a
 * minute apart, which is the whole tolerance being spent here.
 */
export function livePeers(peers: Peer[] | undefined): Peer[] {
  return (peers ?? []).filter((p) => p.seenSecsAgo <= STALE_AFTER_SECS)
}

/** The live peers with `nodeId` open, which is what raises a dot and a banner. */
export function editorsOf(nodeId: string | null, peers: Peer[]): Peer[] {
  if (!nodeId) return []
  return peers.filter((p) => p.editing.includes(nodeId))
}

/**
 * Node id → who has it open, for the navigator.
 *
 * Built once per poll rather than searched per row: a world of several hundred
 * nodes would otherwise scan the peer list once for each of them on every
 * render, to answer a question that is almost always "nobody".
 */
export function editorsByNode(peers: Peer[]): Map<string, string> {
  const byNode = new Map<string, Peer[]>()
  for (const p of peers) {
    for (const id of p.editing) {
      const at = byNode.get(id)
      if (at) at.push(p)
      else byNode.set(id, [p])
    }
  }
  return new Map([...byNode].map(([id, who]) => [id, names(who)]))
}

/**
 * "Nadia", "Nadia and Tomas", "Nadia, Tomas and 2 others".
 *
 * Deduplicated, because the same person with the project open on a desktop and
 * a laptop is two sessions and one name, and "Nadia and Nadia" reads as a bug.
 * The session count in the status bar is the number that stays honest about the
 * difference.
 */
function names(peers: Peer[]): string {
  const unique = [...new Set(peers.map((p) => p.user))]
  const [first, second] = unique
  if (!first) return ''
  if (!second) return first
  if (unique.length === 2) return `${first} and ${second}`
  const rest = unique.length - 2
  return `${first}, ${second} and ${rest} other${rest === 1 ? '' : 's'}`
}

/** Singular against the deduplicated names, not against the session count. */
function verb(peers: Peer[]): string {
  return new Set(peers.map((p) => p.user)).size === 1 ? 'has' : 'have'
}

/**
 * The greeting on open, or `null` when nobody else is here.
 *
 * `null` rather than "nobody else has this open": an empty folder is the
 * ordinary case, and announcing it every time would train people to dismiss the
 * one message that matters.
 */
export function openedText(peers: Peer[]): string | null {
  if (peers.length === 0) return null
  return `${names(peers)} ${verb(peers)} this project open.`
}

/**
 * The passive banner over the editor.
 *
 * It says what happens if both of you save, because that is the only thing the
 * user can act on — and what happens is a conflict file rather than a loss,
 * which is the reassurance that makes not blocking defensible.
 */
export function editingText(peers: Peer[], nodeName: string): string {
  return (
    `${names(peers)} ${verb(peers)} “${nodeName}” open in another session. Nothing is locked ` +
    'and nothing here is switched off — if you both save, the later save is parked beside the ' +
    'file as a conflict for you to resolve rather than overwriting anything.'
  )
}

/** The dot's tooltip. A dot with no explanation is just a mark on a row. */
export function editingTitle(who: string): string {
  return `${who} has this node open in another session — nothing is locked`
}

/**
 * The status bar's session count, or `null` when this is the only session.
 *
 * Sessions rather than people, and ours is one of them: the number answers "how
 * many copies of Wobu are in this folder", which is what the heartbeat files
 * actually say and what makes a second dot on a row explicable.
 */
export function sessionsText(peers: Peer[]): string | null {
  if (peers.length === 0) return null
  return `${peers.length + 1} sessions`
}

/** Names and hosts behind the count, for the status bar's tooltip. */
export function sessionsTitle(peers: Peer[]): string {
  return peers.map((p) => `${p.user} (${p.host})`).join(', ')
}

File: src-tauri/crates/wobu-store/tests/presence.rs
//! Presence across a shared folder, from the outside.
//!
//! The unit tests in `presence.rs` cover one session's own lifecycle. What is
//! exercised here is the part that only exists because there is more than one
//! machine: files written by somebody else's Wobu, on somebody else's clock,
//! read back through ours.
//!
//! Peer session files are written as literal JSON rather than by serialising a
//! `Session`. A round trip would agree with itself no matter what another
//! build actually writes into the folder, and the two things these tests need
//! to vary — the timestamp *inside* the file and the mtime the filesystem
//! reports *for* it — have to be set independently to mean anything.

use std::fs;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime};

use chrono::{Duration as Offset, Utc};
use wobu_core::NodeKind;
use wobu_store::presence::{Presence, STALE_AFTER, peers_in};
use wobu_store::{Project, SaveOutcome};

/// A folder that looks like a project to everything presence touches.
fn folder() -> tempfile::TempDir {
    let dir = tempfile::tempdir().unwrap();
    fs::write(dir.path().join("project.json"), "{}").unwrap();
    fs::create_dir_all(dir.path().join(".wobu/sessions")).unwrap();
    dir
}

fn sessions(root: &Path) -> PathBuf {
    root.join(".wobu").join("sessions")
}

/// Another machine's session file: `heartbeat_at` is what it claims about
/// itself, `written_ago` is what our filesystem will say about it.
fn plant(root: &Path, user: &str, heartbeat_at: &str, written_ago: Duration) -> PathBuf {
    let id = wobu_core::new_id();
    let path = sessions(root).join(format!("{id}.json"));
    fs::write(
        &path,
        format!(
            r#"{{
  "user": "{user}",
  "host": "nadia-mbp",
  "openedAt": "2026-07-31T09:00:00Z",
  "heartbeatAt": "{heartbeat_at}",
  "editing": []
}}"#
        ),
    )
    .unwrap();

    let f = fs::OpenOptions::new().write(true).open(&path).unwrap();
    f.set_modified(SystemTime::now() - written_ago).unwrap();
    path
}

fn stamp(offset: Offset) -> String {
    (Utc::now() + offset).to_rfc3339()
}

#[test]
fn a_session_whose_own_clock_is_an_hour_out_is_still_judged_live_by_its_mtime() {
    // The regression this exists for, and the reason the reaper never reads
    // `heartbeatAt`. Two desktops on a LAN sit minutes or hours apart all the
    // time — a wrong timezone, a VM that resumed, a machine that has never had
    // NTP. Against a sixty-second window that is not a rounding error: a peer
    // an hour behind would be reaped continuously while they sit there typing,
    // and a peer an hour ahead would never expire at all.
    let dir = folder();
    let behind = plant(dir.path(), "nadia", &stamp(-Offset::hours(1)), Duration::from_secs(1));
    let ahead = plant(dir.path(), "sam", &stamp(Offset::hours(1)), Duration::from_secs(1));

    let peers = peers_in(dir.path());
    let mut users: Vec<&str> = peers.iter().map(|p| p.user.as_str()).collect();
    users.sort_unstable();
    assert_eq!(users, ["nadia", "sam"], "a skewed clock decided who was alive");

    assert!(behind.is_file(), "the peer an hour behind was reaped");
    assert!(ahead.is_file(), "the peer an hour ahead was reaped");
    // Both were written a second ago by our own filesystem, and that is the
    // only clock in the answer.
    assert!(peers.iter().all(|p| p.seen_secs_ago <= 2), "{peers:?}");
}

#[test]
fn a_session_that_stopped_beating_is_reaped_however_fresh_it_claims_to_be() {
    // The other half, so the test above cannot pass by never reaping anything.
    // A crashed peer whose last heartbeat happened to carry a timestamp from a
    // clock running fast would otherwise sit in the folder forever.
    let dir = folder();
    let dead =
        plant(dir.path(), "nadia", &stamp(Offset::zero()), STALE_AFTER + Duration::from_secs(30));

    assert!(peers_in(dir.path()).is_empty(), "a session that stopped beating stayed listed");
    assert!(!dead.exists(), "the stale file was left to accumulate");
}

#[test]
fn how_long_ago_a_peer_was_seen_comes_from_the_file_not_from_the_file_contents() {
    // `seenSecsAgo` is what #17 renders and what a stale-presence banner would
    // key off. Deriving it from `heartbeatAt` would put another machine's clock
    // error straight on screen, and nothing on the frontend could tell.
    let dir = folder();
    plant(dir.path(), "nadia", &stamp(Offset::zero()), Duration::from_secs(30));

    let peers = peers_in(dir.path());
    assert_eq!(peers.len(), 1);
    assert!(
        (29..=32).contains(&peers[0].seen_secs_ago),
        "expected roughly 30s, got {}",
        peers[0].seen_secs_ago,
    );
}

#[test]
fn two_sessions_in_one_folder_see_each_other_and_not_themselves() {
    let dir = folder();
    let nadia = Presence::start(dir.path());
    let sam = Presence::start(dir.path());

    let seen_by_nadia = nadia.peers();
    assert_eq!(seen_by_nadia.len(), 1, "{seen_by_nadia:?}");
    assert_eq!(seen_by_nadia[0].session_id, sam.session_id());

    let seen_by_sam = sam.peers();
    assert_eq!(seen_by_sam.len(), 1, "{seen_by_sam:?}");
    assert_eq!(seen_by_sam[0].session_id, nadia.session_id());

    assert_eq!(peers_in(dir.path()).len(), 2, "both are in the folder");
}

#[test]
fn a_session_that_closes_stops_being_a_peer_immediately() {
    // Without the delete on close, leaving a project would leave the person
    // shown as present for another minute — long enough that the next person
    // in decides the feature is lying.
    let dir = folder();
    let nadia = Presence::start(dir.path());
    let sam = Presence::start(dir.path());
    assert_eq!(nadia.peers().len(), 1);

    drop(sam);
    assert!(nadia.peers().is_empty(), "a cleanly closed session lingered");
}

#[test]
fn which_nodes_someone_has_open_crosses_the_folder() {
    let dir = folder();
    let nadia = Presence::start(dir.path());
    let sam = Presence::start(dir.path());

    let kael = wobu_core::new_id();
    sam.set_editing(vec![kael]);

    let peers = nadia.peers();
    assert_eq!(peers.len(), 1);
    assert_eq!(peers[0].editing, vec![kael], "the editing list did not reach the other side");

    // And moving on releases nothing, because nothing was held.
    sam.set_editing(Vec::new());
    assert!(nadia.peers()[0].editing.is_empty());
}

#[test]
fn presence_never_stands_between_anyone_and_a_save() {
    // The line the whole feature is on the right side of. Declaring a node as
    // being edited is a courtesy to the other person, not a claim on the file:
    // hard locks over a share strand files whenever a laptop sleeps, and the
    // recovery is worse than the collision. If this test ever needs a
    // `--force`, something has misread `docs/07-file-shares.md`.
    let dir = tempfile::tempdir().unwrap();
    let mut project = Project::create(dir.path(), "Ashfall").unwrap();
    let kael = project.create_node(NodeKind::Character, "Kael Vantris", None).unwrap();

    let nadia = Presence::start(project.root());
    nadia.set_editing(vec![kael.id]);

    let mut mine = project.get_node(kael.id).unwrap();
    mine.notes_raw = "saved anyway".into();
    let outcome = project.save_node(mine).unwrap();
    assert!(matches!(outcome, SaveOutcome::Saved(_)), "{outcome:?}");
    assert_eq!(project.get_node(kael.id).unwrap().notes_raw, "saved anyway");

    // Deleting it is not blocked either — the dot is information, not a veto.
    project.delete_node(kael.id).unwrap();
}

#[test]
fn opening_a_project_clears_out_what_a_crash_left_behind() {
    // Reaping has to happen without the UI asking. A team that never opens the
    // presence panel still accumulates one file per crash, per closed lid, per
    // dropped VPN, and the listing presence depends on gets slower every month.
    let dir = folder();
    let dead =
        plant(dir.path(), "nadia", &stamp(Offset::zero()), STALE_AFTER + Duration::from_secs(5));
    let live = plant(dir.path(), "sam", &stamp(Offset::zero()), Duration::from_secs(2));

    let _presence = Presence::start(dir.path());

    assert!(!dead.exists(), "opening the project left the dead session behind");
    assert!(live.is_file(), "opening the project reaped a live one");
}

#[test]
fn files_that_are_not_ours_are_left_alone() {
    // A project folder on a share collects other people's tooling, and this
    // reaper deletes things. It only ever deletes what it can name.
    let dir = folder();
    let notes = sessions(dir.path()).join("notes.json");
    let readme = sessions(dir.path()).join("README.txt");
    for path in [&notes, &readme] {
        fs::write(path, "not a session").unwrap();
        let f = fs::OpenOptions::new().write(true).open(path).unwrap();
        f.set_modified(SystemTime::now() - (STALE_AFTER + Duration::from_secs(60))).unwrap();
    }

    assert!(peers_in(dir.path()).is_empty());
    assert!(notes.is_file(), "deleted a file that was never a session");
    assert!(readme.is_file(), "deleted a file that was never a session");
}

/// The one name, in the two places it appears (#76).
///
/// Presence says who is *here* and a conflict sibling says who *wrote* the
/// paragraph you are reading. If those disagree, a user looking at "Nadia is
/// editing Kael Vantris" beside `kael-vantris.conflict-<somebody-else>-….md`
/// cannot tell whether they are the same person, and the whole point of naming
/// a peer is gone. They come from `wobu_store::peer::alias`, and this is the
/// assertion that keeps them coming from the same call.
#[test]
fn presence_and_a_conflict_sibling_call_this_installation_the_same_thing() {
    let dir = tempfile::tempdir().unwrap();
    let mut project = Project::create(dir.path(), "Ashfall").unwrap();
    let node = project.create_node(NodeKind::Character, "Kael Vantris", None).unwrap();

    // Somebody else's Wobu saves over the file while we are holding it.
    let rel = project.index().rel_path_of(node.id).unwrap().expect("indexed");
    let path = wobu_store::paths::from_rel_string(project.root(), &rel);
    let mut ours = project.get_node(node.id).unwrap();
    let theirs = fs::read_to_string(&path).unwrap();
    let (frontmatter, _) = theirs.split_once("\n---\n").unwrap();
    fs::write(&path, format!("{frontmatter}\n---\n\n## Notes\n\nnadia got here first\n")).unwrap();
    ours.notes_raw = "our only copy of this paragraph".into();
    let SaveOutcome::Conflict { conflict_path } = project.save_node(ours).unwrap() else {
        panic!("the fixture needs a conflict and did not get one")
    };

    let _presence = Presence::start(project.root());
    let announced = peers_in(project.root());
    assert_eq!(announced.len(), 1, "{announced:?}");

    let sibling = conflict_path.rsplit('/').next().unwrap();
    let wrote_it = wobu_store::conflict::parse(sibling).expect("a sibling").peer;
    assert_eq!(wrote_it.as_deref(), Some(announced[0].user.as_str()), "{sibling}");
    assert_eq!(wrote_it.as_deref(), Some(wobu_store::peer::alias()));
}
